thunder/dynamo/utils.py

```diff
@@ -831,3 +860,137 @@ def get_torch_compile_kwargs(**kwargs) -> dict:
+
+
+def _torch_inductor_compile(gm, example_inputs) -> torch.nn.GraphModule:
+    class InductorModuleWrapper(torch.nn.Module):
+        def __init__(self, optimized_callable):
+            super().__init__()
+            self.optimized_callable = optimized_callable
+
+        def forward(self, *args):
+            return self.optimized_callable(*args)
+
+    return InductorModuleWrapper(torch._inductor.compile(gm, example_inputs)), "inductor"
+
+
+class ThunderAoTOptimizer:
+    """
+    Helper class that keeps track of profiling data used by the Ahead-of-Time (AoT) optimization process.
+
+    This class maintains mappings between graph module IDs and their corresponding:
+    - dispatch functions (dispatch_map)
+    - original graph modules (id_to_gm_map)
+    - profiling statistics (id_to_profile_stats)
+
+    It also tracks whether profiling is currently active via the is_profiling flag.
+    """
+
+    def __init__(self):
+        self.is_profiling = True
+        self.dispatch_map: dict = {}
+        self.id_to_gm_map: dict = {}
+        self.id_to_profile_stats = {}
+
+
+def default_filter(fn: Callable) -> set[int]:
+    """
+    Default filter function that selects which FX graphs to optimize based on profiling data.
+
+    This function examines the FX graphs collected during profiling and selects only those
+    with static shapes (non-symbolic inputs) for optimization.
+
+    Args:
+        fn: The profiling callable containing collected statistics
+
+    Returns:
+        A set of graph IDs that should be optimized
+    """
+    from torch._inductor.utils import is_symbolic
+
+    id_to_gm = fn._tao.id_to_gm_map
+
+    choosen = set()
+    for idx, gm in id_to_gm.items():
+        dynamic = False
+        placeholders = [n for n in gm.graph.nodes if n.op == "placeholder"]
+        for placeholder in placeholders:
+            example_value = placeholder.meta.get("example_value", None)
+            if example_value is None:
+                continue
+
+            if is_symbolic(example_value):
+                dynamic = True
+                break
+        if not dynamic:
+            choosen.add(idx)
+    return choosen
+
+
+def default_optimizer(gm: torch.fx.GraphModule, stats: ProfileStats) -> Callable:
+    """
+    Default optimizer function that optimizes a GraphModule based on profiling statistics.
+
+    This function:
+    1. Splits the GraphModule into Thunder-supported and unsupported partitions
+    2. Compiles unsupported partitions using PyTorch's Inductor backend
+    3. For supported partitions, selects the best compiler (Thunder or Inductor) based on benchmark results
+    4. Returns the optimized composite GraphModule with all partitions compiled
+
+    Args:
+        gm: The GraphModule to optimize
+        stats: ProfileStats object containing profiling information for the GraphModule
+
+    Returns:
+        The optimized GraphModule
+    """
+    from thunder import jit
```

Comment:
This is where we could add the check that the graph is static vs dynamic, I think